### PR TITLE
Verify challenges

### PR DESCRIPTION
**Pull request by istankovic**
Final part of #1646, this PR adds the `beamer check verify-challenges` command.

Coverage:
image

### CODE REVIEW

File: docs/source/commands.rst [hunk 1]
```diff
@@ -299,3 +300,30 @@ done for all chain pairs, no additional transactions will be made.
...
+
+
+.. _command-check-verify-challenges:
+
+``beamer check verify-challenges``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``beamer check verify-challenges [OPTIONS] FILE``
+
+The command verifies each that challenge stored in ``FILE`` was resolved correctly
```
Comment (bilbeyt) [inline docs/source/commands.rst:312]: just ``each challenge stored in ...``

**Review by bilbeyt — APPROVED** [review]
Generally looking good!

File: docs/source/commands.rst (hunk 1 above)
Comment (istankovic) [reply docs/source/commands.rst]: :eyes:  Fixed, thanks!

File: beamer/check/commands.py [hunk 2]
```diff
@@ -723,3 +723,85 @@ def initiate_challenges(
...
+                fill_chain=challenge.fill_chain,
+            )
+            errors += 1
+            continue
+
+        # Check whether L1 resolution was performed correctly.
+        request_manager = _obtain_request_manager(ctx, challenge.request_chain)
+        claim = request_manager.functions.claims(challenge.claim_id).call()
+        request = request_manager.functions.requests(challenge.request_id).call()
+        if request.filler == claim.claimer and request.fillId == claim.fillId:
```
[the remark this answers is not in the document]
Comment (fredo) [reply beamer/check/commands.py:785]: IOW, it's not sufficient to test whether the agent successfully L1 resolved but also withdrew based on the claim / outcome of the L1 resolution

**Review by fredo — CHANGES_REQUESTED** [review]
Pls Check Comment

File: beamer/check/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (istankovic) [reply beamer/check/commands.py:785]: > IOW, it's not sufficient to test whether the agent successfully L1 resolved but also withdrew based on the claim / outcome of the L1 resolution

Hm, I guess the main question here is what we are testing. If we're testing L1 resolution, which ends once the fields `fillId` and `filler` are set on the request, then it is sufficient to stop once those fields are set. The fact that the current agent immediately withdraws is just a property of its implementation (which we discussed before, in the context of potential changes). 

To be more precise, let me sketch the two possible options for testing:

1. agent's fill + agent's claim + agent invoking the relayer + trusted call chain
2. agent's fill + agent's claim + agent invoking the relayer + trusted call chain + agent's withdraw

We currently have 1, as don't see a real need to also check for the withdraw. But I can also add a check for withdraw, I'm just curious as to why you consider the withdraw behaviour important in this case (as it's totally up to the agent implementation and does not influence the result of the challenge game).

**Comment by bilbeyt** [comment]
as docs changed, you should adapt your doc change

**Comment by istankovic** [comment]
> as docs changed, you should adapt your doc change

Yeah, will rebase, thanks for the note. :+1:

File: beamer/check/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fredo) [reply beamer/check/commands.py:785]: In the past, testing that the agent withdrew based on an L1 resolution was part of the challenge tests. But I agree that you have a valid point here. The command in itself is agent implementation agnostic, and could be reused for other agent implementations. Those may not withdraw the tokens immediately. 

On the other hand, I look at what we really want to test. And here testing that the agent did not get the tokens out can be a crucial element. As I see it, we want to test that the agent actually wins the challenge by performing the L1 resolution. So IMO the whole context of testing is to win the challenge and not only to invoke the relayer. 
There could be a scenario where the agent, due to a bug, calls the relayer too late. The L1 resolution would be performed but the challenge could have been already resolved. In this scenario option 1) would still succeed, option 2) however would not. 
I would argue that this is what we want to test.

[the remark this answers is not in the document]
Comment (istankovic) [reply beamer/check/commands.py:785]: > There could be a scenario where the agent, due to a bug, calls the relayer too late. The L1 resolution would be performed but the challenge could have been already resolved. In this scenario option 1) would still succeed, option 2) however would not.

Right, that's a good point. Thanks for clarification, will amend the check to also consider withdraw.

[the remark this answers is not in the document]
Comment (istankovic) [reply beamer/check/commands.py]: Done.

File: beamer/check/commands.py [hunk 3]
```diff
@@ -723,3 +723,89 @@ def initiate_challenges(
...
+            )
+            errors += 1
+            continue
+
+        # Check whether L1 resolution was performed correctly.
+        request_manager = _obtain_request_manager(ctx, challenge.request_chain)
+        claim = request_manager.functions.claims(challenge.claim_id).call()
+        request = request_manager.functions.requests(challenge.request_id).call()
+        if (
+            request.withdrawClaimId == challenge.claim_id
```
Comment (fredo) [inline beamer/check/commands.py:786]: Unfortunately, this does improve the situation only a bit. It makes sure that it was withdrawn by this very claimId but not before or after the resolution, or IOW it doesn't state who the beneficiary was. However this is not so easy to find out. So I would be fine with this tradeoff

**Review by fredo — APPROVED** [review]
Please read my last comment, if you have any good idea how to solve it easily, then feel free to fix otherwise i'm fine as is

File: beamer/check/commands.py (hunk 3 above)
Comment (istankovic) [reply beamer/check/commands.py:786]: > It makes sure that it was withdrawn by this very claimId but not before or after the resolution

Actually, due to the fact that `request.filler` is only set by `resolveRequest`, it's guaranteed that L1 resolution occurred if that condition is true. However, ....

> IOW it doesn't state who the beneficiary was.

... is definitely true, we don't have an easy way to check that the beneficiary was the claimer.
The only thing that comes to my mind is looking for the exact withdrawal event and comparing the addresses, which would be :woozy_face: .

Here I'm not so much concerned about the complexity of event-based checking, we have all the infrastructure to make that relatively straightforward. Rather, I'm more concerned about the overall "cost" of the command i.e. now verification is quite quick and costs nothing and when I weigh that against the event-based checking... eh, I think I prefer more the quick-and-cheap-to-verify (also in presence of many challenges in one file).

**Pull request merged**